packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -2557,462 +2136,151 @@ class _RenderCupertinoDialogActions extends RenderBox
-      final bool isButtonPressed = currentButtonParentData.isPressed;
-
-      bool isPrevButtonPressed = false;
-      if (prevChild != null) {
-        assert(prevChild.parentData is _ActionButtonParentData);
-        final _ActionButtonParentData previousButtonParentData = prevChild.parentData! as _ActionButtonParentData;
-        isPrevButtonPressed = previousButtonParentData.isPressed;
+  bool _debugHasValidConstraints(BoxConstraints constraints) {
+    FlutterError? constraintsError;
+    ErrorSummary? errorSummary;
+    assert(() {
+      if (constraints.maxWidth == double.infinity) {
+        errorSummary = ErrorSummary('The incoming width constraints are unbounded.');
       }
-
-      final bool isDividerPresent = child != firstChild;
-      final bool isDividerPainted = isDividerPresent && !(isButtonPressed || isPrevButtonPressed);
-      final Rect dividerRect = Rect.fromLTWH(
-        accumulatingOffset.dx,
-        accumulatingOffset.dy,
-        size.width,
-        dividerThickness,
-      );
-
-      final Rect buttonBackgroundRect = Rect.fromLTWH(
-        accumulatingOffset.dx,
-        accumulatingOffset.dy + (isDividerPresent ? dividerThickness : 0.0),
-        size.width,
-        child.size.height,
-      );
-
-      // If this button is pressed, then we don't want a white background to be
-      // painted, so we erase this button from the background path.
-      if (isButtonPressed) {
-        backgroundFillPath.addRect(buttonBackgroundRect);
-        pressedBackgroundFillPath.addRect(buttonBackgroundRect);
+      if (errorSummary != null) {
+        constraintsError = FlutterError.fromParts(<DiagnosticsNode>[
+          errorSummary!,
+          ErrorDescription('The incoming constraints are: $constraints'),
+        ]);
       }
-
-      // If this divider is needed, then we erase the divider area from the
-      // background path, and on top of that we paint a translucent gray to
-      // darken the divider area.
-      if (isDividerPainted) {
-        backgroundFillPath.addRect(dividerRect);
-        dividersPath.addRect(dividerRect);
+      return true;
+    }());
+    if (constraintsError != null) {
+      throw constraintsError!;
+    }
+    return true;
+  }
+
+  bool _useHorizontalLayout(double overallWidth) {
+    // Horizontal layout only applies to cases of 3 children: 2 action buttons
+    // and 1 divider.
+    if (childCount != 3) {
+      return false;
+    }
+    final double slotWidth = horizontalSlotWidthFor(overallWidth: overallWidth);
+    RenderBox child = firstChild!;
+    while (true) {
+      // If both children fit into a half-row slot, use the horizontal layout.
+      // Max intrinsic widths are used here, which, according to
+      // [TextPainter.maxIntrinsicWidth], allows text to be displayed at their
+      // full font size.
+      if (child.getMaxIntrinsicWidth(double.infinity) > slotWidth) {
```

Comment:
nit: should this be the actual maxHeight instead of `infinity`? In case there's RenderAspectRatio or RenderWrap in the tree.

Reply:
If I understand correctly there isn't a maxHeight here, because in the worst case, the actions section is laid out in the vertical mode, which scrolls and allows the buttons to be as tall as they want (if they don't restrict themselves as 1 line max.)